### ISSUE TITLE
Use of universal2 wheels in py2app

### ISSUE THREAD

**Issue by mattip** [ronaldoussoren/py2app#399]
Over at NumPy we are discussing the need for universal2 wheels in numpy/numpy#20787. It seems the only consumer of them would be py2app (or other projects that know how to bundle libraries with the universal2 python distribution). Would it be helpful for py2app if NumPy has a universal2 wheel on PyPI? How would you download and use it? The cost to us is extra CI cycles and space on PyPI, is the benefit worth it?

Maybe you could extend the [documentation](https://py2app.readthedocs.io/en/latest/tweaking.html#universal-binaries) to clarify the expectations from library maintainers.

**Comment by ronaldoussoren** [comment]
I definitely need to work on this documention, the bit you link to is old documentation from the PPC to Intel migration.

The current version of py2app expects to see C extensions that are compatible with all architectures that you'd want to run the application on.  For current systems that would be binaries supporting both arm64 and x86_64, and I'd prefer to have binaries were the deployment target matches that of the python.org installers (arm64 11.0+ and x86_64 10.9+).

Py2app cannot merge two architecture specific binaries at the moment, although I'm considering adding that in the future. That will have to wait for time to perform some serious surgery on the code base though.

**Comment by glyph** [comment]
I think this is arguably also a bug in `pip`. https://github.com/pypa/pip/issues/11573

**Comment by joshwilson-dbx** [comment]
I found this after attempting to bundle `numpy` into an app using the `universal2` architecture. As they have declared that they are not going to be generating `universal2` wheels (numpy/numpy#20787), I'll need to find another solution here.

The reason I was even looking at `universal2` support in the first place was because I need to sign my app, which needs to be able to run on both Intel and Apple Silicon. `py2app` will do the signing for me, but only for the `universal2` and `arm64` architectures. Why can it not also sign `x86_64`? I'm fine with doing two builds, one for each architecture, but I'm not sure how best to proceed with regard to the signing my `x86_64` builds.

**Comment by glyph** [comment]
I wrote a little gizmo to ensure universal2 wheels, among other things, including smashing vendor-provided binaries together with delocate-fuse  https://github.com/glyph/Encrust/blob/91b51faf299ad5967d1853a0651c27666107a156/src/encrust/_architecture.py#L165-L167

This project is very much a temporary solution until the community can converge on something more standardized; see also https://github.com/beeware/briefcase/issues/1217

**Comment by joshwilson-dbx** [comment]
That Encrust library looks very interesting, as it's tackling the exact problem I have. Unfortunately I can't use it out-of-the-box due to some peculiarities in my project, but the `delocate-fuse` tool is promising.

As of now, I've manually fused the `numpy` wheels and pipenv installed directly from the generated wheel prior to running `py2app` as normal. That seems to be working. Or at least it let me get to the next failing library :). I see that the Encrust tool will do this discovery and fusing automatically and internally.

FYI: My current flow is to use `py2app` to generate a signed app, then use `pkgbuild --sign...` to generate the signed installer file. I'm not using the notarizing flow (yet). It looks like Encrust runs the whole gamut. (Also, it seems to replicate the `codesign` functionality already present in `py2app` but maybe in a better way?)

(Also, I wanted to say thanks for twisted. That's one amazing and useful library.)

**Comment by glyph** [comment]
> That Encrust library looks very interesting, as it's tackling the exact problem I have. Unfortunately I can't use it out-of-the-box due to some peculiarities in my project, but the `delocate-fuse` tool is promising.

I'd be thrilled if you could use it as-is, but it exists more as a resource for this sort of thing than as a tool in its own right. It does All The Stuff so you can see what All The Stuff is, and hopefully it won't need to exist at some point :).

(That said, if you want to submit a PR to account for your own peculiarities, I'll do my best to integrate it.)

**Comment by glyph** [comment]
> As of now, I've manually fused the `numpy` wheels and pipenv installed directly from the generated wheel prior to running `py2app` as normal. That seems to be working. Or at least it let me get to the next failing library :). I see that the Encrust tool will do this discovery and fusing automatically and internally.

Yep.  You may be able to invoke just that part, the stages are *mostly* separate internally?

**Comment by glyph** [comment]
> FYI: My current flow is to use `py2app` to generate a signed app, then use `pkgbuild --sign...` to generate the signed installer file. I'm not using the notarizing flow (yet). It looks like Encrust runs the whole gamut. (Also, it seems to replicate the `codesign` functionality already present in `py2app` but maybe in a better way?)

On modern versions of macOS, signing is kind of useless without notarization.  Unless there's some special rule for pkg files?  Or you're submitting to the app store or something?

I don't know that the codesign functionality is *better*, but I've used bits of encrust with app bundles that have come out of non-py2app tools, which is why that code is there.  Also my earliest versions of this probably predate py2app doing signing.

**Comment by glyph** [comment]
> (Also, I wanted to say thanks for twisted. That's one amazing and useful library.)

Thanks for saying so! This feels like a good forum to point out that Twisted has a built-in CoreFoundation reactor, so you can use it within your mac applications if you like :)

**Comment by joshwilson-dbx** [comment]
> > That Encrust library looks very interesting, as it's tackling the exact problem I have. Unfortunately I can't use it out-of-the-box due to some peculiarities in my project, but the `delocate-fuse` tool is promising.
> 
> I'd be thrilled if you could use it as-is, but it exists more as a resource for this sort of thing than as a tool in its own right. It does All The Stuff so you can see what All The Stuff is, and hopefully it won't need to exist at some point :).
> 
> (That said, if you want to submit a PR to account for your own peculiarities, I'll do my best to integrate it.)

So actually, I was able to use `encrust fatten` right out-of-the-box. After it installed the fused wheels in my dev/build machine's virtual environment I was able to build and sign a universal2 app with my normal `py2app` invocation.

I've opened some issues on that project to track what we're discussing here. I didn't want to get too far off topic here ;) . I'd be glad to work on some PRs when I find some time.

As for this issue, I think we have a workaround for those that need universal2 binaries in their `py2app` app but they aren't provided by the library maintainers. Assuming a fused wheel can be build for them, just run `encrust fatten` prior to `py2app`.
